### PR TITLE
EA-1774 registration of api key is prohibited if it already exists

### PR DESCRIPTION
**Pull request by SrishtiSingh-eu**

### CODE REVIEW

File: apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java [hunk 1]
```diff
@@ -142,6 +144,12 @@ public ApikeyController(ApikeyRepo apikeyRepo, CaptchaManager captchaManager, Cu
             LOG.debug(e.getMessage() + ", abort creating apikey");
             return new ResponseEntity<>(e.getMessage(), HttpStatus.BAD_REQUEST);
         }
+        try{
+            apiKeyAlreadyExist(apikeyCreate);
+        } catch (ApikeyException e) {
+            LOG.debug("Api key already exist for " + e.getMessage());
+            return new ResponseEntity<>(e.getMessage(), HttpStatus.BAD_REQUEST);
+        }
```
Comment (P-Ehlert) [inline apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java]: Exceptions should only be thrown if there is a problem from which you can't recover. They should not be used in a 'normal' flow of events. In this case I feel it would be better to let the apiKeyAlreadyExists return a boolean and check that. That is more efficient than throwing and catching an error.

File: apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java [hunk 2]
```diff
@@ -646,9 +661,11 @@ private Apikey copyUpdateValues(Apikey apikey, ApikeyDetails apikeyUpdate){
...
         String retval = "Required parameter(s): ";
         ArrayList<String> missingList = new ArrayList<>();
-        if (null == apikeyUpdate.getFirstName()) missingList.add("'firstName'");
-        if (null == apikeyUpdate.getLastName()) missingList.add("'lastName'");
-        if (null == apikeyUpdate.getEmail()) missingList.add("'email'");
+        if (apikeyUpdate.getFirstName()==null || apikeyUpdate.getFirstName().isEmpty()) missingList.add("'firstName'");
+        if (null == apikeyUpdate.getLastName() || apikeyUpdate.getLastName().isEmpty()) missingList.add("'lastName'");
+        if (null == apikeyUpdate.getEmail() || apikeyUpdate.getEmail().isEmpty()) missingList.add("'email'");
+        if (null == apikeyUpdate.getAppName() || apikeyUpdate.getAppName().isEmpty()) missingList.add("'appName'");
+        if (null == apikeyUpdate.getCompany() || apikeyUpdate.getCompany().isEmpty()) missingList.add("'company'");
```
Comment (P-Ehlert) [inline apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java]: For consistency in code style please use {} with if-statements.

File: apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java [hunk 3]
```diff
@@ -646,9 +661,11 @@ private Apikey copyUpdateValues(Apikey apikey, ApikeyDetails apikeyUpdate){
     private void mandatoryMissing(ApikeyAction apikeyUpdate) throws ApikeyException {
         String retval = "Required parameter(s): ";
         ArrayList<String> missingList = new ArrayList<>();
-        if (null == apikeyUpdate.getFirstName()) missingList.add("'firstName'");
-        if (null == apikeyUpdate.getLastName()) missingList.add("'lastName'");
-        if (null == apikeyUpdate.getEmail()) missingList.add("'email'");
+        if (apikeyUpdate.getFirstName()==null || apikeyUpdate.getFirstName().isEmpty()) missingList.add("'firstName'");
```
Comment (P-Ehlert) [inline apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java]: This is fine, but you could also use if (!StringUtils.isEmpty(apiKeyUpdate.getFirstName()) to achieve the same, or StringUtils.isBlank... to also check if the string contains at least 1 non white-space character

File: apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java [hunk 4]
```diff
@@ -658,6 +675,15 @@ private void mandatoryMissing(ApikeyAction apikeyUpdate) throws ApikeyException
             throw new ApikeyException(400, BAD_EMAIL_FORMAT, BAD_EMAIL_FORMAT);
         }
     }
-
+    private void apiKeyAlreadyExist(ApikeyAction apikeyUpdate) throws ApikeyException{
+        Apikey apikey= this.apikeyRepo.findByEmailAndAppName(apikeyUpdate.getEmail(), apikeyUpdate.getAppName());
+        System.out.println(apikey);
```
Comment (P-Ehlert) [inline apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java]: Please do not use System.out or System.err (unless it's temporary for debugging). For permanent code use LOG.debug instead or LOG.info instead.

File: apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java [hunk 5]
```diff
@@ -658,6 +675,15 @@ private void mandatoryMissing(ApikeyAction apikeyUpdate) throws ApikeyException
...
+        if(apikey!=null){
+            String message = APIKEYALREADYEXIST +" for : " +apikeyUpdate.getEmail() +", " +apikeyUpdate.getAppName();
+            throw new ApikeyException(400, APIKEYALREADYEXIST, message);
```
Comment (P-Ehlert) [inline apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java]: More efficient to just return apikey != null and use that value to send the correct response

File: apikey-service/src/test/java/eu/europeana/apikey/controller/ApikeyControllerTest.java [hunk 6]
```diff
@@ -49,12 +49,12 @@
...
     public void setup()throws Exception {
         Apikey apikey = new Apikey(EXISTING_API_KEY, "edward",
-                "potts", "potts@mail.com");
+                "potts", "potts@mail.com","appNme", "company");
         apikey.setKeycloakId(EXISTING_API_KEY);
         apikeyRepo.saveAndFlush(apikey);
 
         apikey = new Apikey(DEPRECATED_API_KEY, "frank",
-                "sinatra", "sinatra@mail.com");
+                "sinatra", "sinatra@mail.com","appName", "company");
```
Comment (P-Ehlert) [inline apikey-service/src/test/java/eu/europeana/apikey/controller/ApikeyControllerTest.java]: I propose to also add a test case where the appname and company are null, because we have quite a bit of those in our current database
